processor/src/stack/mod.rs

```diff
@@ -4,11 +4,89 @@ use core::cmp;
 // STACK
 // ================================================================================================
 
-/// TODO: add comments
+#[derive(Clone)]
+struct OverflowTrace {
+    added_clk: usize,
+    removed_clk: Option<usize>,
+    value: Felt,
+}
+
+/// Overflow table stores a list of overflow values from the stack.
+/// We create a separate structure to be able to store historical changes
+/// in the overflow table, so we can be able to reconstruct what's in the overflow table
+/// at any given point in time.
+#[derive(Clone)]
+pub struct Overflow {
+    active: Vec<Felt>,
+    trace: Vec<OverflowTrace>,
+    trace_enabled: bool,
+}
+
+impl Overflow {
+    pub fn new() -> Self {
+        Self {
+            active: Vec::new(),
+            trace: Vec::new(),
+            trace_enabled: false,
+        }
+    }
+
+    /// Enabling tracing on the overflow table.
+    pub fn set_trace(&mut self, enabled: bool) {
+        self.trace_enabled = enabled;
+    }
+
+    pub fn push(&mut self, step: usize, value: Felt) {
+        self.active.push(value);
+        if self.trace_enabled {
+            self.trace.push(OverflowTrace {
+                added_clk: step,
+                removed_clk: None,
+                value,
+            })
+        }
+    }
+
+    pub fn pop(&mut self, step: usize) -> Option<Felt> {
+        let popped = self.active.pop();
+        if popped.is_some() && self.trace_enabled {
+            let len = self.trace.len();
+            for i in (0..len).rev() {
+                if self.trace[i].removed_clk.is_none() {
+                    self.trace[i].removed_clk = Some(step);
+                    break;
+                }
+            }
+        }
+
+        popped
+    }
+
+    /// Get values up to a certain step in the overflow table.
+    pub fn get_values_at(&self, step: usize) -> Vec<Felt> {
+        self.trace
+            .iter()
+            .filter(|&x| x.added_clk <= step && x.removed_clk.unwrap_or(usize::MAX) > step)
+            .map(|x| x.value)
+            .collect()
```

Comment:
With this structure, it seems like we'd need to do a full linear scan of the trace to get the current state. This may be OK, but I wonder if we could do anything better than this.

Reply:
Ok I changed this to stop at a specific clock cycle when scanning.